[Clang] Implement P2718R0 "Lifetime extension in range-based for loops"

yronglin opened the pull request:
Implement P2718R0 "Lifetime extension in range-based for loops" (https://wg21.link/P2718R0)

Differential Revision: https://reviews.llvm.org/D153701

yronglin commented:
ping~

AaronBallman commented on clang/lib/Parse/ParseStmt.cpp:
```diff
@@ -2288,10 +2288,11 @@ StmtResult Parser::ParseForStatement(SourceLocation *TrailingElseLoc) {
 
-  // Similarly, we need to do the semantic analysis for a for-range
-  // statement immediately in order to close over temporaries correctly.
+    // Similarly, we need to do the semantic analysis for a for-range
+    // statement immediately in order to close over temporaries correctly.
   } else if (ForEach) {
     ForEachStmt = Actions.ActOnObjCForCollectionStmt(ForLoc,
```
```suggestion
  } else if (ForEach) {
    // Similarly, we need to do the semantic analysis for a for-range
    // statement immediately in order to close over temporaries correctly.
    ForEachStmt = Actions.ActOnObjCForCollectionStmt(ForLoc,
```
Might as well move the comment to where it belongs.

yronglin replied:
Agree, done.

AaronBallman commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -6289,9 +6289,19 @@ ExprResult Sema::BuildCXXDefaultArgExpr(SourceLocation CallLoc,
+      if (V.HasImmediateCalls)
+        ExprEvalContexts.back().DelayedDefaultInitializationContext = {
+            CallLoc, Param, CurContext};
+      // Pass down lifetime extendning flag, and collect temporaries in
```
```suggestion
      // Pass down lifetime extending flag, and collect temporaries in
```

yronglin replied:
fixed

AaronBallman commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -6305,6 +6315,10 @@ ExprResult Sema::BuildCXXDefaultArgExpr(SourceLocation CallLoc,
         return ExprError();
       Init = Res.get();
+
+      // Collect MaterializeTemporaryExprs in the rewrited CXXDefaultArgExpr.
```
```suggestion
      // Collect MaterializeTemporaryExprs in the rewritten CXXDefaultArgExpr.
```

yronglin replied:
fixed

AaronBallman commented on clang/lib/Sema/SemaExprCXX.cpp:
```diff
@@ -8244,11 +8229,33 @@ ExprResult Sema::IgnoredValueConversions(Expr *E) {
+    // able to synthesize the storage for itself in the aggregate case, and
+    // adding the extra node to the AST is just clutter.
+    if (ShouldMaterializePRValueInDiscardedExpression() &&
+        getLangOpts().CPlusPlus17 && E->isPRValue() &&
```
Can you explain why the check for C++17?

AaronBallman replied:
Ah okay, so this is still related to the lifetime extension for range-based for loops -- we needed this bit for C++17 conformance which we previously didn't have to do?

(Mostly trying to reason about whether this is a separate bug fix that should be landed in a separate patch or whether this is related to the correctness for P2718R0.)

yronglin replied:
Yeah, I think so. we previously didn't materialize temporary object may should also be in compliance with the standard words "the materialization of a temporary object is generally delayed as long as possible in order to avoid creating unnecessary temporary object".

AaronBallman commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -2539,6 +2539,12 @@ StmtResult Sema::ActOnCXXForRangeStmt(Scope *S, SourceLocation ForLoc,
   }
 
+  if (getLangOpts().CPlusPlus23) {
+    auto Entity = InitializedEntity::InitializeVariable(RangeVar);
```
Please spell out the type instead of using `auto`. Also, should this be a const ref to avoid a copy?

yronglin replied:
I use`InitializedEntity ` instead of auto, `InitializedEntity::InitializeVariable` returns a rvalue, so I haven't add a const ref.

AaronBallman replied:
Ah `InitializedEntity` doesn't have a copy constructor so the move operation will be cheap. Thanks!

AaronBallman commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -2783,6 +2789,7 @@ StmtResult Sema::BuildCXXForRangeStmt(SourceLocation ForLoc,
       LoopVar->setType(SubstAutoTypeDependent(LoopVar->getType()));
     }
   } else if (!BeginDeclStmt.get()) {
+
```
Spurious whitespace change can be removed.

yronglin replied:
removed

AaronBallman reviewed (commented):
Thank you for this! I think it looks pretty close, but I'd love to get more eyes on this given the importance of the feature. CC @zygoloid @rjmccall @erichkeane

yronglin commented:
Add `ReleaseNotes`.

yronglin commented:
Rebase to fix CI issue

cor3ntin commented:
@zygoloid Can you take another crack at this? If you have time to confirm the design maybe we could try to land it in Clang 18. Otherwise the plan is to be conservative. Thanks!

zygoloid commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -9959,6 +9976,18 @@ class Sema final {
+    return ExprEvalContexts.back().IsInLifetimeExtendingContext;
+  }
+
+  bool ShouldMaterializePRValueInDiscardedExpression() const {
```
Is there a reason that this is capitalized and the function above is not? (Genuine question, I've lost track of LLVM's preferred capitalization rule for functions, but this locally looks inconsistent.)

yronglin replied:
LLVM prefer `shouldMaterializePRValueInDiscardedExpression` https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly , I'll rename it.

zygoloid commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -1357,6 +1363,19 @@ class Sema final {
     // VLAs).
     bool InConditionallyConstantEvaluateContext = false;
 
+    /// Whether we are currently in a context in which temporaries must be
+    /// lifetime-extended (Eg. in a for-range initializer).
+    bool IsInLifetimeExtendingContext = false;
```
I think we should be clear here that this is only for unusual lifetime-extension contexts, and not just (for example) in an initializer of a reference.
```suggestion
    /// Whether we are currently in a context in which all temporaries must be
    /// lifetime-extended, even if they're not bound to a reference (for example,
    /// in a for-range initializer).
    bool IsInLifetimeExtendingContext = false;
```

zygoloid reviewed (commented):
This looks good to me.

shafik commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -2312,12 +2312,31 @@ Parser::DeclGroupPtrTy Parser::ParseDeclGroup(ParsingDeclSpec &DS,
         FRI->RangeExpr = ParseBraceInitializer();
       else
         FRI->RangeExpr = ParseExpression();
+      if (getLangOpts().CPlusPlus23)
```
Why not move this into the `if (getLangOpts().CPlusPlus23)` above?

cor3ntin replied:
`ForRangeLifetimeExtendTemps` is going to be filled by `ParseExpression`/ I wonder if we should do that unconditionally though.
and replace the if with `assert(getLangOpts().CPlusPlus23 || Actions.ExprEvalContexts.back().ForRangeLifetimeExtendTemps.empty())`

yronglin replied:
agree

shafik commented on clang/test/CXX/special/class.temporary/p6.cpp:
```diff
@@ -238,3 +239,236 @@ void init_capture_init_list() {
   // CHECK: }
 }
+
+namespace P2718R0 {
```
Is it possible to also test the case where it does not clean up like in the proposal:

```cpp
for (auto e : f2(g())) {}  // undefined behavior
```

yronglin replied:
agree, I have compiled this case, Currently, clang will not extending lifetime of temporaries in this case.

cor3ntin commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -1338,6 +1338,12 @@ class Sema final {
     llvm::SmallPtrSet<DeclRefExpr *, 4> ReferenceToConsteval;
 
+    /// P2718R0 - Lifetime extension in range-based for loops.
+    /// MaterializeTemporaryExprs in for-range-init expression which need to
```
```suggestion
    /// MaterializeTemporaryExprs in for-range-init expressions which need to
```

cor3ntin commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -6291,9 +6291,19 @@ ExprResult Sema::BuildCXXDefaultArgExpr(SourceLocation CallLoc,
+      // CreateMaterializeTemporaryExpr when we rewrite the call argument.
+      auto &LastRecord = ExprEvalContexts.back();
+      auto &PrevRecord = ExprEvalContexts[ExprEvalContexts.size() - 2];
+      LastRecord.IsInLifetimeExtendingContext = IsInLifetimeExtendingContext;
+
```
We probably should do the same thing for `BuildCXXDefaultInitExpr`, right?

```cpp

struct my_range {
  // ...
  foo bar = temp();
};

for(elem : my_range {}) {}
```

yronglin replied:
Good catch! Yeah, we need do the same thing for `BuildCXXDefaultInitExpr`, I'm working on it.

cor3ntin commented:
@yronglin Do you know when you will be able to get to this? thanks!

yronglin commented:
> @yronglin Do you know when you will be able to get to this? thanks!

Sorry for the lack of new actions for a long time. I have recently fixed some problems in the dependent context. I will try to complete the DefaultInit related functions and tests this weekend.

yronglin commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -1361,6 +1367,20 @@ class Sema final {
+    ///
+    /// We do not materialize temporaries by default in order to avoid creating
+    /// unnecessary temporary objects.
+    bool MaterializePRValueInDiscardedExpression = false;
```
I think this variable and related function `needMaterializePRValueInDiscardedExpression()` should be renamed to  `NeedMaterializePRValue` and `needMaterializePRValue()` , then it can use in other cases.

In [class.temporary P2][http://eel.is/c++draft/class.temporary#2]

The materialization of a temporary object is generally delayed as long as possible in order to avoid creating unnecessary temporary objects[.](http://eel.is/c++draft/class.temporary#2.sentence-1)
[Note [3](http://eel.is/c++draft/class.temporary#note-3): Temporary objects are materialized:
[(2.1)](http://eel.is/c++draft/class.temporary#2.1)
when binding a reference to a prvalue ([[dcl.init.ref]](http://eel.is/c++draft/dcl.init.ref), [[expr.type.conv]](http://eel.is/c++draft/expr.type.conv), [[expr.dynamic.cast]](http://eel.is/c++draft/expr.dynamic.cast), [[expr.static.cast]](http://eel.is/c++draft/expr.static.cast), [[expr.const.cast]](http://eel.is/c++draft/expr.const.cast), [[expr.cast]](http://eel.is/c++draft/expr.cast)),
[(2.2)](http://eel.is/c++draft/class.temporary#2.2)
when performing member access on a class prvalue ([[expr.ref]](http://eel.is/c++draft/expr.ref), [[expr.mptr.oper]](http://eel.is/c++draft/expr.mptr.oper)),
[(2.3)](http://eel.is/c++draft/class.temporary#2.3)
when performing an array-to-pointer conversion or subscripting on an array prvalue ([[conv.array]](http://eel.is/c++draft/conv.array), [[expr.sub]](http://eel.is/c++draft/expr.sub)),
[(2.4)](http://eel.is/c++draft/class.temporary#2.4)
when initializing an object of type std​::​initializer_list<T> from a [braced-init-list](http://eel.is/c++draft/dcl.init.general#nt:braced-init-list) ([[dcl.init.list]](http://eel.is/c++draft/dcl.init.list)),
[(2.5)](http://eel.is/c++draft/class.temporary#2.5)
for certain unevaluated operands ([[expr.typeid]](http://eel.is/c++draft/expr.typeid), [[expr.sizeof]](http://eel.is/c++draft/expr.sizeof)), and
[(2.6)](http://eel.is/c++draft/class.temporary#2.6)
when a prvalue that has type other than cv void appears as a discarded-value expression ([[expr.context]](http://eel.is/c++draft/expr.context))[.](http://eel.is/c++draft/class.temporary#2.sentence-2)
— end note]

yronglin committed:
[Clang] Implement P2718R0 "Lifetime extension in range-based for loops"

 Implement P2718R0 "Lifetime extension in range-based for loops" (https://wg21.link/P2718R0)

Differential Revision: https://reviews.llvm.org/D153701

yronglin commented:
Should we also dump the sub-expr which in `CXXDefaultArgExpr`:
for example:
Consider:
```
struct A {
  A();
  ~A();
};

typedef int vec3[3];
vec3 &f(const A &a = A{});

void foo() {
  for (auto e : f()) {}
}
```
Current AST Dump:
```
FunctionDecl 0x126813170 <line:85:1, line:87:1> line:85:6 foo 'void ()'
  `-CompoundStmt 0x126814720 <col:12, line:87:1>
    `-CXXForRangeStmt 0x1268146b0 <line:86:3, col:23>
      |-<<<NULL>>>
      |-DeclStmt 0x1268137b8 <col:17>
      | `-VarDecl 0x126813548 <col:17, col:19> col:17 implicit used __range1 'vec3 &' cinit
      |   `-ExprWithCleanups 0x126813730 <col:17, col:19> 'vec3':'int[3]' lvalue
      |     `-CallExpr 0x126813318 <col:17, col:19> 'vec3':'int[3]' lvalue
      |       |-ImplicitCastExpr 0x126813300 <col:17> 'vec3 &(*)(const A &)' <FunctionToPointerDecay>
      |       | `-DeclRefExpr 0x126813278 <col:17> 'vec3 &(const A &)' lvalue Function 0x126813030 'f' 'vec3 &(const A &)'
      |       `-CXXDefaultArgExpr 0x1268133f0 <<invalid sloc>> 'const A' lvalue
      |-DeclStmt 0x126814448 <col:15>
      | `-VarDecl 0x126813880 <col:15> col:15 implicit used __begin1 'int *' cinit
 .......
```
We can't see directly on the AST whether the lifetime is extended or not. The sub-expr which in `CXXDefaultArgExpr` is:
```
ExprWithCleanups 0x1398c55d8 'const A':'const struct A' lvalue
`-MaterializeTemporaryExpr 0x1398c55c0 'const A':'const struct A' lvalue extended by Var 0x1398c5748 '__range1' 'vec3 &'
  `-ImplicitCastExpr 0x1398c55a8 'const A':'const struct A' <NoOp>
    `-CXXBindTemporaryExpr 0x1398c5588 'A':'struct A' (CXXTemporary 0x1398c5588)
      `-CXXTemporaryObjectExpr 0x1398c5550 'A':'struct A' 'void (void)'
```

cor3ntin commented:
> Should we also dump the sub-expr which in `CXXDefaultArgExpr`: for example: Consider:
> 
> ```
> struct A {
>   A();
>   ~A();
> };
> 
> typedef int vec3[3];
> vec3 &f(const A &a = A{});
> 
> void foo() {
>   for (auto e : f()) {}
> }
> ```
> 
> Current AST Dump:
> 
> ```
> FunctionDecl 0x126813170 <line:85:1, line:87:1> line:85:6 foo 'void ()'
>   `-CompoundStmt 0x126814720 <col:12, line:87:1>
>     `-CXXForRangeStmt 0x1268146b0 <line:86:3, col:23>
>       |-<<<NULL>>>
>       |-DeclStmt 0x1268137b8 <col:17>
>       | `-VarDecl 0x126813548 <col:17, col:19> col:17 implicit used __range1 'vec3 &' cinit
>       |   `-ExprWithCleanups 0x126813730 <col:17, col:19> 'vec3':'int[3]' lvalue
>       |     `-CallExpr 0x126813318 <col:17, col:19> 'vec3':'int[3]' lvalue
>       |       |-ImplicitCastExpr 0x126813300 <col:17> 'vec3 &(*)(const A &)' <FunctionToPointerDecay>
>       |       | `-DeclRefExpr 0x126813278 <col:17> 'vec3 &(const A &)' lvalue Function 0x126813030 'f' 'vec3 &(const A &)'
>       |       `-CXXDefaultArgExpr 0x1268133f0 <<invalid sloc>> 'const A' lvalue
>       |-DeclStmt 0x126814448 <col:15>
>       | `-VarDecl 0x126813880 <col:15> col:15 implicit used __begin1 'int *' cinit
>  .......
> ```
> 
> We can't see directly on the AST whether the lifetime is extended or not. The sub-expr which in `CXXDefaultArgExpr` is:
> 
> ```
> ExprWithCleanups 0x1398c55d8 'const A':'const struct A' lvalue
> `-MaterializeTemporaryExpr 0x1398c55c0 'const A':'const struct A' lvalue extended by Var 0x1398c5748 '__range1' 'vec3 &'
>   `-ImplicitCastExpr 0x1398c55a8 'const A':'const struct A' <NoOp>
>     `-CXXBindTemporaryExpr 0x1398c5588 'A':'struct A' (CXXTemporary 0x1398c5588)
>       `-CXXTemporaryObjectExpr 0x1398c5550 'A':'struct A' 'void (void)'
> ```

I think this makes sense, but it can be a separate patch, unless it helps for testing.


FYI, we prefer merge commits instead of rebase + force push because Github reviews tend to loose context when doing force pushes / rebases in PRs

cor3ntin commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -6375,12 +6383,16 @@ ExprResult Sema::BuildCXXDefaultInitExpr(SourceLocation Loc, FieldDecl *Field) {
+    // Pass down lifetime extending flag, and collect temporaries in
+    // CreateMaterializeTemporaryExpr when we rewrite the call argument.
+    keepInLifetimeExtendingContext();
+    keepInMaterializeTemporaryObjectContext();
```
Where are the tests for BuildCXXDefaultInitExpr?

yronglin replied:
Thanks for your review! I still have some questions about member init.
For this case(https://godbolt.org/z/jPKe8M6xE), should we materialize the temporaries in class Z::x's default init,  but unless I'm mistaken?

yronglin replied:
@cor3ntin I tried this code https://godbolt.org/z/n77v9rMTz , and the temporaries which in `CXXDefaultInitExpr`  has been lifetime extended in AST. But the LLVM IR unexpected, and the lifetime of the temporaries was not extended. Could you please guide me as to why this is and whether I have made some mistake?

cor3ntin replied:
I think part of the problem might be that we should only extend in the aggregate case.
ie, when we call `BuildCXXDefaultInit` from `FillInEmptyInitForField`.

In the constructor case, the init lifetime is bound to that of the member initializer expression.

yronglin replied:
Thanks for your help! I have a case use aggregate initialization (https://godbolt.org/z/vn8Y11ja8). I think neither the brace-or-equals initializer nor default member initializer should not be materialized and lifetime-extended in this code snippet, seems they don't satisfy any of [class.temporary P2](http://eel.is/c++draft/class.temporary#2) . @hubert-reinterpretcast Could you please give me an example where `CXXDefaultInitExpr` has to be lifetime-extended? I saw you comment in D153701 that we need to do lifetime-extend for `CXXDefaultInitExpr`. I'm sorry that I'm unfamiliar with that part of clang and I have no clue how to write tests for CXXDefaultInitExpr.

cor3ntin replied:
I think you are right. 
I'm sorry, that whole thing is a bit confusing to me.

But the initializer is used in the construction of a subobject, which is different from a default argument which is initialized on call site... so I think the conclusion is that we should do nothing for default inits, I think

yronglin replied:
Thank you for your consideration.  I think so too, whether I have made some mistake, I will try to ask the author of P2718R0

hubert-reinterpretcast replied:
> @hubert-reinterpretcast Could you please give me an example where `CXXDefaultInitExpr` has to be lifetime-extended?

https://github.com/llvm/llvm-project/issues/85613

yronglin replied:
Many thanks for your information.

yronglin replied:
I try to extend lifetime of the `MaterializedTemporaryExpr`s in `CXXDefaultInitExpr`, but looking from the generated IR, `MaterializedTemporaryExpr ::setExtendingDecl` does not take effect. I think some modifications may be needed in CodeGen.

yronglin commented:
For this case, should

> > Should we also dump the sub-expr which in `CXXDefaultArgExpr`: for example: Consider:
> 
> I think this makes sense, but it can be a separate patch, unless it helps for testing.

Cool! Let's do it in a separate patch.

> FYI, we prefer merge commits instead of rebase + force push because Github reviews tend to loose context when doing force pushes / rebases in PRs

Thanks for the reminder, I may not have converted from the Phabricator habit yet. I tried to rebase to fix a libcxx ci issue(gdb_pretty_printer_test)

cor3ntin commented:
Here is what I'd like to do @yronglin
Revert the changes to BuildCXXDefaultInit.
And land it in main.
Then we'll have a few weeks to decide if we backport it to 18. Dies that sound good to you?

cor3ntin commented on clang/www/cxx_status.html:
```diff
@@ -429,7 +429,7 @@ <h2 id="cxx23">C++23 implementation status</h2>
       <td>Lifetime extension in range-based for loops</td>
       <td><a href="https://wg21.link/P2718R0">P2718R0</a></td>
-      <td class="none" align="center">No</td>
+      <td class="none" align="center">Clang 18</td>
```
```suggestion
      <td class="none" align="center">Clang 19</td>
```

yronglin replied:
done

yronglin committed:
Revert changes of CXXDefaultInitExpr

Signed-off-by: yronglin <yronglin777@gmail.com>

yronglin commented:
> Here is what I'd like to do @yronglin Revert the changes to BuildCXXDefaultInit. And land it in main. Then we'll have a few weeks to decide if we backport it to 18. Dies that sound good to you?

Yeah, it's looks good to me, I have reverted changes for `CXXDefaultInitExpr`.

yronglin committed:
Revert whitespace change

Signed-off-by: yronglin <yronglin777@gmail.com>

cor3ntin reviewed (approved):
LGTM. 
Do you need me to merge that for you?

yronglin commented:
> LGTM. Do you need me to merge that for you?

Thanks a lot for your review and help! I have commit access, once the CI green, I'll land this patch.

yronglin committed:
Merge branch 'main' into arcpatch-D153701

yronglin committed:
Update cxx_status.html

Signed-off-by: yronglin <yronglin777@gmail.com>
